### ISSUE TITLE
AIX PPC 64 - Includes are incorrect, and methods do not exist.

### ISSUE THREAD

**Issue by bemehiser** [protocolbuffers/protobuf#4531]
When building on AIX PPC, the wrong header file is included.

This was produced compiling `protobuf 3.5.1.1` with `gcc 4.8.3` on `AIX 6.1`

[AIX definition](https://github.com/google/protobuf/blob/9497a657d577ebda0272711651c3dcdda3a4ac35/src/google/protobuf/stubs/platform_macros.h#L112)
```
#ifdef _AIX
#define GOOGLE_PROTOBUF_OS_AIX
```

Because `GOOGLE_PROTOBUF_OS_AIX` is defined, `google/protobuf/stubs/atomicops_internals_power.h` is [included](https://github.com/google/protobuf/blob/860bd12fec5c69e6529565165532b3d5108a7d97/src/google/protobuf/stubs/atomicops.h#L189)
```
#elif defined(GOOGLE_PROTOBUF_OS_AIX)
#include <google/protobuf/stubs/atomicops_internals_power.h>
```

I'm building with ppc gcc, so I expect `google/protobuf/stubs/atomicops_internals_ppc_gcc.h` to be included. I can see that if arch ppc, `atomicops_internals_ppc_gcc.h` should be included, but it doesn't seem to be.
```
#elif defined(GOOGLE_PROTOBUF_ARCH_PPC)
#include <google/protobuf/stubs/atomicops_internals_ppc_gcc.h>
```

When I change the `GOOGLE_PROTOBUF_OS_AIX` include to include `atomicops_internals_ppc_gcc.h`, protobuf builds, but fails to link protoc.

```
[ 99%] Linking CXX executable protoc
/aspera/build/atc-1.2/aix-6.1-ppc32/bin/cmake -E cmake_link_script CMakeFiles/protoc.dir/link.txt --verbose=1
/usr/bin/c++      CMakeFiles/protoc.dir/src/google/protobuf/compiler/main.cc.o  -o protoc -Wl,-brtl,-bnoipath,-bexpall libprotobuf.a libprotoc.a libprotobuf.a -lpthreads /usr/lib/libz.a -Wl,-blibpath:/opt/freeware/lib/gcc/powerpc-ibm-aix6.1.0.0/4.8.3:/opt/freeware/lib:/usr/lib:/lib
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::Acquire_Load(long const volatile*)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::Acquire_CompareAndSwap(long volatile*, long, long)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::Release_Store(long volatile*, long)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::NoBarrier_AtomicIncrement(long volatile*, long)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::NoBarrier_Store(long volatile*, long)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::NoBarrier_Load(long const volatile*)
ld: 0711-317 ERROR: Undefined symbol: .google::protobuf::internal::Release_CompareAndSwap(long volatile*, long, long)
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
gmake[2]: *** [protoc] Error 1
gmake[2]: Leaving directory `/aspera/usr/bemehiser/protobuf'
gmake[1]: *** [CMakeFiles/protoc.dir/all] Error 2
gmake[1]: Leaving directory `/aspera/usr/bemehiser/protobuf'
gmake: *** [all] Error 2
```

When I check for one of the symbols, I can see that it's in `libprotobuf.a`, which is what I expect. The symbol should have been defined in `google/protobuf/stubs/atomicops_internals_ppc_gcc.h`.

```
# nm libprotoc.a | grep Acquire_Load
# ...
._ZN6google8protobuf8internal12Acquire_LoadEPVKl U           -
._ZN6google8protobuf8internal12Acquire_LoadEPVKi T        1548
._ZN6google8protobuf8internal12Acquire_LoadEPVKl U           -
_ZN6google8protobuf8internal12Acquire_LoadEPVKi D      116772          12
._ZN6google8protobuf8internal12Acquire_LoadEPVKl U           -
# ...
```

**Comment by xfxyjwf** [comment]
Could you help send us a pull request to make the issue?

**Comment by bemehiser** [comment]
The symbol was caused because the 64 bit functions were never defined in `atomicops_internals_ppc_gcc.h`.

Adding the following code solved the compilation error. Note that I don't know PowerPC 64 assembly, so I just commented out those sections. If anyone knows assembly, I assume these snippets would be fairly simple to rebuild for 64 bit PowerPC.

```diff
Index: atomicops_internals_ppc_gcc.h
===================================================================
@@ -146,6 +146,87 @@
   return *ptr;
 }

+#ifdef GOOGLE_PROTOBUF_ARCH_64_BIT
+
+inline Atomic64 NoBarrier_CompareAndSwap(volatile Atomic64 *ptr,
+                                         Atomic64 old_value,
+                                         Atomic64 new_value) {
+    Atomic64 prev;
+
+//    __asm__ __volatile__(
+//    "0:                                  \n\t"
+//    "lwarx %[prev],0,%[ptr]              \n\t"
+//    "cmpw 0,%[prev],%[old_value]         \n\t"
+//    "bne- 1f                             \n\t"
+//    "stwcx. %[new_value],0,%[ptr]        \n\t"
+//    "bne- 0b                             \n\t"
+//    "1:                                  \n\t"
+//    : [prev] "=&r"(prev), "+m"(*ptr)
+//    : [ptr] "r"(ptr), [old_value] "r"(old_value), [new_value] "r"(new_value)
+//    : "cc", "memory");
+
+    return prev;
+}
+
+
+inline Atomic64 NoBarrier_AtomicIncrement(volatile Atomic64 *ptr,
+                                          Atomic64 increment) {
+  Atomic64 temp;
+
+//  __asm__ __volatile__(
+//      "0:                                  \n\t"
+//      "lwarx %[temp],0,%[ptr]              \n\t"
+//      "add %[temp],%[increment],%[temp]    \n\t"
+//      "stwcx. %[temp],0,%[ptr]             \n\t"
+//      "bne- 0b                             \n\t"
+//      : [temp] "=&r"(temp)
+//      : [increment] "r"(increment), [ptr] "r"(ptr)
+//      : "cc", "memory");
+
+  return temp;
+}
+
+inline Atomic64 Barrier_AtomicIncrement(volatile Atomic64 *ptr,
+                                        Atomic64 increment) {
+    MemoryBarrierInternal();
+    Atomic64 res = NoBarrier_AtomicIncrement(ptr, increment);
+    MemoryBarrierInternal();
+    return res;
+}
+
+inline Atomic64 Acquire_CompareAndSwap(volatile Atomic64 *ptr,
+                                       Atomic64 old_value, Atomic64 new_value) {
+  Atomic64 res = NoBarrier_CompareAndSwap(ptr, old_value, new_value);
+  MemoryBarrierInternal();
+  return res;
+}
+
+inline Atomic64 Release_CompareAndSwap(volatile Atomic64 *ptr,
+                                       Atomic64 old_value, Atomic64 new_value) {
+  MemoryBarrierInternal();
+  Atomic64 res = NoBarrier_CompareAndSwap(ptr, old_value, new_value);
+  return res;
+}
+
+inline void NoBarrier_Store(volatile Atomic64 *ptr, Atomic64 value) {
+  *ptr = value;
+}
+
+inline void Release_Store(volatile Atomic64 *ptr, Atomic64 value) {
+  MemoryBarrierInternal();
+  *ptr = value;
+}
+
+inline Atomic64 NoBarrier_Load(volatile const Atomic64 *ptr) { return *ptr; }
+
+inline Atomic64 Acquire_Load(volatile const Atomic64 *ptr) {
+    Atomic64 value = *ptr;
+    MemoryBarrierInternal();
+    return value;
+}
+
+#endif
+
 }  // namespace internal
 }  // namespace protobuf
 }  // namespace google
```

**Comment by bemehiser** [comment]
I tried to write 64 bit methods. The problem is that I couldn't have local labels, because for some reason the directional branching wasn't working (`bne- 0b:`). I tried using named labels, but with compiler optimization I got duplicate symbol errors because the methods were inlined. When I disabled compiler optimization, I got massive symbol tables which caused the linker to fail.

With the named labels, and code optimizations off, I did manage to get the code to compile sometimes (it still failed to link randomly), but I don't think it was good enough to merge.

If anyone knows ppc64 assembly, this should be an easy fix.
